pyro/distributions/relaxed_straight_through.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+from pyro.distributions.torch import RelaxedOneHotCategorical
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(RelaxedOneHotCategorical)
+class RelaxedCategoricalStraightThrough(RelaxedOneHotCategorical):
+    """
+    Implementation of RelaxedCategorical with a Straight-through Estimator
+
+    This distribution allows using a Gumbel-Softmax thing to estimate probabilities for samples but generates discrete samples.
+    In detail, it uses discrete samples for the forward step but the relaxed sample for the backwards gradient.
+    """
+    def __init__(self, temperature, probs=None, logits=None, validate_args=None):
+        super(RelaxedCategoricalStraightThrough, self).__init__(temperature=temperature, probs=probs, logits=logits, validate_args=validate_args)
+        self._unquantize = {}
+
+    def rsample(self, sample_shape=torch.Size()):
+        soft_sample = super(RelaxedCategoricalStraightThrough, self).rsample(sample_shape)
+        hard_sample = QuantizeCategorical.apply(soft_sample)
+        self._unquantize[id(hard_sample)] = soft_sample
```

Comment:
why not use the tensor itself as key?

Reply:
ok, the hard_sample is a onehot vector mapping of a dense vector.
However, multiple dense vectors can map to the same onehot vector.

As such, in order to make the mapping unique, we decided to use the id instead of the value.

Makes sense?